File: mobile_agent/lib/utils/assignment_pick.dart
import '../models/assignment.dart';
import 'time_utils.dart';

bool _sameCalendarDate(DateTime a, DateTime b) =>
    a.year == b.year && a.month == b.month && a.day == b.day;

/// Même marge qu'au serveur (`checkins.window.END_GRACE_AFTER_MINUTES`).
const kEndGraceMinutes = 120;

/// True si l'heure locale actuelle tombe dans le créneau [a], en tenant compte
/// de [shiftDate] (indispensable : sans cela, le poste d'hier 06h–18h paraît
/// « actif » le lendemain matin au même horaire).
bool assignmentIsActiveNow(Assignment a, DateTime now) {
  final today = DateTime(now.year, now.month, now.day);
  final yesterday = today.subtract(const Duration(days: 1));
  final nowMin = now.hour * 60 + now.minute;
  final startMin = minutesFromHhMm(a.startTime);
  final endMin = minutesFromHhMm(a.endTime);
  final crossesMidnight = startMin > endMin;
  final endWithGrace = endMin + kEndGraceMinutes;

  if (!crossesMidnight) {
    if (!_sameCalendarDate(a.shiftDate, today)) return false;
    return nowMin >= startMin && nowMin < endWithGrace;
  }
  if (_sameCalendarDate(a.shiftDate, today) && nowMin >= startMin) return true;
  // Matin après minuit : jusqu'à fin + marge (ex. nuit 18h→06h, clôture jusqu'à ~08h).
  if (_sameCalendarDate(a.shiftDate, yesterday) && nowMin < endWithGrace) {
    return true;
  }
  return false;
}

/// Poste à afficher : d'abord celui déjà commencé sans fin (clôture en cours),
/// puis celui actif maintenant, sinon le premier du jour civil local, sinon secours.
Assignment? pickActiveAssignment(List<Assignment> list, DateTime now) {
  if (list.isEmpty) return null;
  for (final a in list) {
    if (a.hasStart && !a.hasEnd) return a;
  }
  for (final a in list) {
    if (assignmentIsActiveNow(a, now)) return a;
  }
  final today = DateTime(now.year, now.month, now.day);
  for (final a in list) {
    if (_sameCalendarDate(a.shiftDate, today)) return a;
  }
  return list.first;
}

/// Choisit le poste affiché après sync API.
/// Ne réapplique pas [preserveAssignmentId] s'il existe déjà un poste commencé
/// sans fin (évite d'écraser la nuit d'hier par l'affectation du jour).
Assignment? resolveSelectedAssignment(
  List<Assignment> list,
  DateTime now, {
  int? preserveAssignmentId,
}) {
  final resolved = pickActiveAssignment(list, now);
  if (preserveAssignmentId == null) return resolved;
  final hasOpenShift = list.any((a) => a.hasStart && !a.hasEnd);
  if (hasOpenShift) return resolved;
  for (final a in list) {
    if (a.id == preserveAssignmentId) return a;
  }
  return resolved;
}

File: backend/shifts/assignment_pick.py
"""Sélection de l'affectation active pour un vigile (aligné sur l'app mobile)."""

from __future__ import annotations

from datetime import date, datetime, time, timedelta

from django.utils import timezone

from shifts.models import ShiftAssignment


def _same_calendar_date(a: date, b: date) -> bool:
    return a.year == b.year and a.month == b.month and a.day == b.day


def _minutes_from_time(t: time) -> int:
    return t.hour * 60 + t.minute


def assignment_is_active_now(assignment: ShiftAssignment, now: datetime) -> bool:
    local = timezone.localtime(now)
    today = local.date()
    yesterday = today - timedelta(days=1)
    now_min = local.hour * 60 + local.minute
    start_min = _minutes_from_time(assignment.start_time)
    end_min = _minutes_from_time(assignment.end_time)
    crosses_midnight = start_min > end_min

    if not crosses_midnight:
        if not _same_calendar_date(assignment.shift_date, today):
            return False
        return start_min <= now_min < end_min
    if _same_calendar_date(assignment.shift_date, today) and now_min >= start_min:
        return True
    if _same_calendar_date(assignment.shift_date, yesterday) and now_min < end_min:
        return True
    return False


def pick_active_assignment(
    assignments: list[ShiftAssignment],
    *,
    now: datetime | None = None,
) -> ShiftAssignment | None:
    if not assignments:
        return None
    now = now or timezone.now()
    for row in assignments:
        if assignment_is_active_now(row, now):
            return row
    today = timezone.localdate()
    for row in assignments:
        if row.shift_date == today:
            return row
    return assignments[0]


def assignments_for_guard(
    guard_id: int,
    *,
    site_id: int | None = None,
    now: datetime | None = None,
) -> list[ShiftAssignment]:
    now = now or timezone.now()
    today = timezone.localdate()
    yesterday = today - timedelta(days=1)
    qs = ShiftAssignment.objects.select_related("site", "guard").filter(
        guard_id=guard_id,
        shift_date__in=[yesterday, today],
        status__in=ShiftAssignment.active_on_duty_statuses(),
    )
    if site_id is not None:
        qs = qs.filter(site_id=site_id)
    return list(qs.order_by("shift_date", "start_time"))


def pick_assignment_for_guard(
    guard_id: int,
    *,
    site_id: int | None = None,
    now: datetime | None = None,
) -> ShiftAssignment | None:
    return pick_active_assignment(
        assignments_for_guard(guard_id, site_id=site_id, now=now),
        now=now,
    )
